File: deploy/docker/canvasbench/context/continue/continue-wrapper.py
#!/usr/bin/env python3
import hashlib
import json
import os
from pathlib import Path
import shutil
import subprocess
import sys
import tempfile
from urllib.parse import urlsplit


base_url = os.environ.get("OPENAI_BASE_URL", "").rstrip("/")
api_key = os.environ.get("OPENAI_API_KEY", "")
if not base_url or not api_key:
    raise SystemExit("canvasbench Continue contract requires OPENAI_BASE_URL and OPENAI_API_KEY")
parsed = urlsplit(base_url)
if parsed.scheme not in {"http", "https"} or not parsed.netloc or parsed.path.rstrip("/") != "/v1":
    raise SystemExit("canvasbench Continue contract requires a valid OpenAI /v1 base URL")

state = Path(tempfile.mkdtemp(prefix="canvasbench-continue-", dir="/tmp"))
try:
    config_path = state / "config.yaml"
    config = {
        "name": "CanvasBench", "version": "1.0.0", "schema": "v1",
        "models": [{
            "name": "canvasbench", "provider": "openai",
            "model": os.environ.get("CANVASBENCH_ADAPTER_MODEL", ""),
            "apiBase": base_url, "apiKey": "${{ secrets.OPENAI_API_KEY }}",
            "contextLength": int(os.environ.get("CANVASBENCH_CONTEXT_WINDOW", "32768")),
            "defaultCompletionOptions": {
                "maxTokens": int(os.environ.get("CANVASBENCH_MAX_OUTPUT_TOKENS", "4096")),
                "temperature": float(os.environ.get("CANVASBENCH_TEMPERATURE", "0")),
                "topP": float(os.environ.get("CANVASBENCH_TOP_P", "1")),
            },
            "capabilities": ["tool_use"], "roles": ["chat", "edit", "apply"],
        }],
    }
    config_path.write_text(json.dumps(config, separators=(",", ":")) + "\n")
    config_path.chmod(0o600)
    args = [str(config_path) if value == "__CANVASBENCH_CONFIG__" else value for value in sys.argv[1:]]
    completed = subprocess.run(
        ["node", "/opt/harness/cn-real.js", *args], env=dict(os.environ), text=True,
        stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
    )
    output = completed.stdout.strip() or f"Continue exited {completed.returncode} without output"
    session = hashlib.sha256("\0".join(sys.argv[1:]).encode()).hexdigest()[:24]
    print(json.dumps({
        "schema": "canvasbench.cli-wrapper.v1", "harness": "continue",
        "session_id": f"continue-{session}", "output": output, "exit_code": completed.returncode,
    }, separators=(",", ":")))
    raise SystemExit(completed.returncode)
finally:
    shutil.rmtree(state, ignore_errors=True)

File: deploy/docker/canvasbench/context/continue/continue-wrapper.mjs
#!/usr/bin/env node
import crypto from "node:crypto";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { spawnSync } from "node:child_process";

const baseURL = (process.env.OPENAI_BASE_URL || "").replace(/\/+$/, "");
const apiKey = process.env.OPENAI_API_KEY || "";
if (!baseURL || !apiKey) throw new Error("canvasbench Continue contract requires OPENAI_BASE_URL and OPENAI_API_KEY");
const parsed = new URL(baseURL);
if (!["http:", "https:"].includes(parsed.protocol) || parsed.pathname.replace(/\/+$/, "") !== "/v1") {
  throw new Error("canvasbench Continue contract requires a valid OpenAI /v1 base URL");
}

const state = fs.mkdtempSync(path.join(os.tmpdir(), "canvasbench-continue-"));
try {
  const configPath = path.join(state, "config.yaml");
  const config = {
    name: "CanvasBench",
    version: "1.0.0",
    schema: "v1",
    models: [{
      name: "canvasbench",
      provider: "openai",
      model: process.env.CANVASBENCH_ADAPTER_MODEL || "",
      apiBase: baseURL,
      apiKey: "${{ secrets.OPENAI_API_KEY }}",
      contextLength: Number(process.env.CANVASBENCH_CONTEXT_WINDOW || 32768),
      defaultCompletionOptions: {
        maxTokens: Number(process.env.CANVASBENCH_MAX_OUTPUT_TOKENS || 4096),
        temperature: Number(process.env.CANVASBENCH_TEMPERATURE || 0),
        topP: Number(process.env.CANVASBENCH_TOP_P || 1),
      },
      capabilities: ["tool_use"],
      roles: ["chat", "edit", "apply"],
    }],
  };
  fs.writeFileSync(configPath, `${JSON.stringify(config)}\n`, { mode: 0o600 });
  const originalArgs = process.argv.slice(2);
  const args = originalArgs.map((value) => value === "__CANVASBENCH_CONFIG__" ? configPath : value);
  const completed = spawnSync(process.execPath, ["/opt/harness/cn-real.js", ...args], {
    env: process.env,
    encoding: "utf8",
    maxBuffer: 32 * 1024 * 1024,
  });
  if (completed.error) throw completed.error;
  const exitCode = completed.status ?? 1;
  const captured = `${completed.stdout || ""}${completed.stderr || ""}`.trim();
  const output = captured || `Continue exited ${exitCode} without output`;
  const session = crypto.createHash("sha256").update(originalArgs.join("\0")).digest("hex").slice(0, 24);
  process.stdout.write(`${JSON.stringify({
    schema: "canvasbench.cli-wrapper.v1",
    harness: "continue",
    session_id: `continue-${session}`,
    output,
    exit_code: exitCode,
  })}\n`);
  process.exitCode = exitCode;
} finally {
  fs.rmSync(state, { recursive: true, force: true });
}
